File: it-agents/skills/code-architect/references/code-examples/cpp/State.cpp
// State — let an object alter behavior when its state changes (state objects).
#include <iostream>
#include <memory>
#include <string>

class Machine;

// State interface.
class State {
public:
    virtual ~State() = default;
    virtual void push(Machine& m) = 0;
    virtual std::string name() const = 0;
};

class Machine {
    std::unique_ptr<State> state;
public:
    explicit Machine(std::unique_ptr<State> initial) : state(std::move(initial)) {}
    void setState(std::unique_ptr<State> s) { state = std::move(s); }
    void push() { state->push(*this); }
    std::string current() const { return state->name(); }
};

// Forward declarations resolved below.
class Idle; class Running; class Paused;

class Idle : public State {
public:
    void push(Machine& m) override;
    std::string name() const override { return "Idle"; }
};
class Running : public State {
public:
    void push(Machine& m) override;
    std::string name() const override { return "Running"; }
};
class Paused : public State {
public:
    void push(Machine& m) override;
    std::string name() const override { return "Paused"; }
};

void Idle::push(Machine& m)    { m.setState(std::make_unique<Running>()); }
void Running::push(Machine& m) { m.setState(std::make_unique<Paused>()); }
void Paused::push(Machine& m)  { m.setState(std::make_unique<Idle>()); }

int main() {
    Machine machine{std::make_unique<Idle>()};
    for (int i = 0; i < 4; ++i) {
        std::cout << "state: " << machine.current() << "\n";
        machine.push();
    }
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/rust/state.rs
// State — let an object change its behavior when its internal state changes, by
// delegating to interchangeable state objects that return the next state.

trait State {
    // Consumes self and returns the next state.
    fn next(self: Box<Self>) -> Box<dyn State>;
    fn label(&self) -> &'static str;
}

struct Idle;
struct Running;
struct Done;

impl State for Idle {
    fn next(self: Box<Self>) -> Box<dyn State> {
        Box::new(Running)
    }
    fn label(&self) -> &'static str {
        "Idle"
    }
}
impl State for Running {
    fn next(self: Box<Self>) -> Box<dyn State> {
        Box::new(Done)
    }
    fn label(&self) -> &'static str {
        "Running"
    }
}
impl State for Done {
    fn next(self: Box<Self>) -> Box<dyn State> {
        self // terminal state stays put
    }
    fn label(&self) -> &'static str {
        "Done"
    }
}

struct Machine {
    state: Box<dyn State>,
}

impl Machine {
    fn step(&mut self) {
        let current = std::mem::replace(&mut self.state, Box::new(Idle));
        self.state = current.next();
    }
}

fn main() {
    let mut machine = Machine { state: Box::new(Idle) };
    for _ in 0..4 {
        println!("state: {}", machine.state.label());
        machine.step();
    }
}
